[Fix #4896] Fix Style/DateTime wrongly triggered on classes `...::DateTime`

dpostorivo opened the pull request:
This fixes #4896 by checking if DateTime is in another namespace.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests(`rake spec`) are passing.
* [x] The new code doesn't generate RuboCop offenses that are checked by `rake internal_investigation`.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [ ] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://chris.beams.io/posts/git-commit/

Drenmi commented:
I'm gonna have to ask not to use regular expressions on the source code for this. It's very frail, and is causing us issues in other cops already. Please check the structure of the inspected AST instead. Thanks! 🙂

dpostorivo commented:
@Drenmi I tried to do it using the information in the node, but I couldn't find a fool proof method. Do you have any example of looking for namespace inside rubocop I could use as an example?

Drenmi commented:
@dpostorivo: Sure! 🙂 

We can use the `ruby-parse` command line utility to inspect the AST for a few different cases.

The cases that should probably be considered offensive are: `DateTime` and `::DateTime`, while `Foo::DateTime` should be left alone.

```
```

pocke commented:
I guess we need a predicate like `top_level_const?`(and I think we should implement it in another pull-request.).

```
(const top_level_const? :DateTime)
```

dpostorivo commented:
@Drenmi thanks for the help and for the example of how to use ruby-parse. I just updated the change to include what you discussed and added a test for the cbase example.

bbatsov commented:
Btw, won't this still yield false positive within modules defining some `DateTime` class and referring to it unqualified? I'm guessing this will rarely happen in practice, but it should be kept in mind.

bbatsov commented:
Anyways, I'll merge this as an improvement over the existing situation.
